File: server/quests/global/spells/26508.pl
# Marked Passage (Item 26508) - Expedition Compatible
# First use: Marks location and teleports to Guild Lobby
# Second use: Returns to marked location (including expedition instances)
# Author: Straps

sub EVENT_SPELL_EFFECT_CLIENT {

    my $guild_lobby_zone_id = 344; # Guild Lobby
    my $bucket_key = "marked_passage_" . $client->CharacterID();
    my $bucket_ttl = 1800; # 30 minutes

    # Check if we have a stored location
    my $stored_location = quest::get_data($bucket_key);

    if ($stored_location) {
        # Second use: Return to stored location
        my @coords = split(/,/, $stored_location);

        # Now expecting 6 values: zone_id, instance_id, x, y, z, heading
        if (scalar(@coords) == 6) {
            my ($zone_id, $instance_id, $x, $y, $z, $heading) = @coords;

            # Delete the stored location
            quest::delete_data($bucket_key);

            # Store teleport data in a timer-specific bucket
            my $teleport_key = "marked_passage_teleport_" . $client->CharacterID();
            quest::set_data($teleport_key, $stored_location, 5); # 5 second TTL

            # Set a timer to perform the actual teleport
            quest::settimer("marked_passage_return", 1);

            $client->Message(10, "Returning to marked location...");
        } else {
            $client->Message(13, "Error: Invalid stored location data.");
        }

    } else {
        # First use: Store current location and port to Guild Lobby
        my $current_zone_id = $zoneid;
        my $current_instance_id = $client->GetInstanceID();
        my $x = $client->GetX();
        my $y = $client->GetY();
        my $z = $client->GetZ();
        my $heading = $client->GetHeading();

        # Store location in data bucket (now with instance_id)
        my $location_data = "$current_zone_id,$current_instance_id,$x,$y,$z,$heading";
        quest::set_data($bucket_key, $location_data, $bucket_ttl);

        # Random spawn location in Guild Lobby
        my @guild_lobby_spawns = (
            [   0,  315,  2, 0 ],   # center-ish, commonly safe
            [ -262, 414,  2, 0 ],   # west side, clear floor
            [  257, 411,  2, 0 ],   # east side, clear floor
            [   -2, 515,  2, 0 ],   # north side, clear floor
        );

        # Pick random spawn point
        my $random_spawn = $guild_lobby_spawns[int(rand(scalar(@guild_lobby_spawns)))];
        my $spawn_data = join(",", $guild_lobby_zone_id, 0, @$random_spawn); # Instance 0 for Guild Lobby

        # Store teleport destination in timer-specific bucket
        my $teleport_key = "marked_passage_teleport_" . $client->CharacterID();
        quest::set_data($teleport_key, $spawn_data, 5); # 5 second TTL

        # Set a timer to perform the actual teleport
        quest::settimer("marked_passage_goto", 1);

        $client->Message(10, "Location marked. Teleporting to Guild Lobby...");
    }

}

File: server/quests/global/spells/26508.lua
-- Marked Passage AA (spell 26508)
-- First use: marks current zone and position
-- Second use: teleports to the marked position and clears it
function event_spell_effect(e)
    local client = eq.get_entity_list():GetClientByID(e.caster_id)
    if not client or not client.valid then return end

    local stored_zone = client:GetGlobal("mp_zone")

    if stored_zone == "" then
        client:SetGlobal("mp_zone", tostring(client:GetZoneID()), 5, "F")
        client:SetGlobal("mp_x",    tostring(client:GetX()),       5, "F")
        client:SetGlobal("mp_y",    tostring(client:GetY()),       5, "F")
        client:SetGlobal("mp_z",    tostring(client:GetZ()),       5, "F")
        client:SetGlobal("mp_h",    tostring(client:GetHeading()), 5, "F")
        client:Message(15, "Your location has been marked. Use Marked Passage again to return here.")
    else
        local zone_id = tonumber(stored_zone)
        local x       = tonumber(client:GetGlobal("mp_x"))
        local y       = tonumber(client:GetGlobal("mp_y"))
        local z       = tonumber(client:GetGlobal("mp_z"))
        local heading = tonumber(client:GetGlobal("mp_h"))

        client:DeleteGlobal("mp_zone")
        client:DeleteGlobal("mp_x")
        client:DeleteGlobal("mp_y")
        client:DeleteGlobal("mp_z")
        client:DeleteGlobal("mp_h")

        client:Message(15, "You are transported to your marked location.")
        client:MovePC(zone_id, x, y, z, heading)
    end
end
